internal/xds/balancer/clusterimpl/tests/balancer_test.go

```diff
@@ -663,3 +663,201 @@ func (s) TestLRSLogicalDNS(t *testing.T) {
+	}
+
+	// Start two server backends exposing the test service.
+	backend1 := stubserver.StartTestService(t, nil)
+	defer backend1.Stop()
+	backend1Host, backend1Port := hostAndPortFromAddress(t, backend1.Address)
+
+	backend2 := stubserver.StartTestService(t, nil)
+	defer backend2.Stop()
+	backend2Host, backend2Port := hostAndPortFromAddress(t, backend2.Address)
+
+	// Configure the xDS management server with Logical DNS cluster pointing to backend1.
+	const serviceName = "test-xds-service"
+	resources := e2e.DefaultClientResources(e2e.ResourceParams{
+		DialTarget: serviceName,
+		NodeID:     nodeID,
+		Host:       backend1Host,
+		Port:       backend1Port,
+	})
+	resources.Clusters = []*v3clusterpb.Cluster{
+		e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+			ClusterName: "cluster-" + serviceName,
+			Type:        e2e.ClusterTypeLogicalDNS,
+			DNSHostName: backend1Host,
+			DNSPort:     backend1Port,
+		}),
+	}
+	resources.Endpoints = nil
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a ClientConn and make a successful RPC.
+	cc, err := grpc.NewClient(fmt.Sprintf("xds:///%s", serviceName), grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithResolvers(resolverBuilder),
+	)
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	cc.Connect()
+	client := testgrpc.NewTestServiceClient(cc)
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
+
+	// Perform initial RPC to backend1, expect success.
+	var peer peer.Peer
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer)); err != nil {
+		t.Fatalf("Expected initial rpc to backend1 to succeed, got: %v", err)
+	}
+
+	// Verify that the request was sent to backend 1.
+	if got, want := peer.Addr.String(), backend1.Address; got != want {
+		t.Fatalf("EmptyCall() routed to backend %q, want %q", peer.Addr, backend1.Address)
+	}
+
+	// Stop backend1 to simulate failure.
+	backend1.Stop()
+
+	// Update management server with Logical DNS cluster pointing to backend2 now.
+	resources.Clusters[0] = e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: "cluster-" + serviceName,
+		Type:        e2e.ClusterTypeLogicalDNS,
+		DNSHostName: backend2Host,
+		DNSPort:     backend2Port,
+	})
+
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("failed to update management server resources to backend2: %v", err)
+	}
+
+	// Wait for client to re-resolve and start routing to backend2.
+	for {
+		if ctx.Err() != nil {
+			t.Fatalf("Timed out while waiting for RPC to be routed to backend2: %v", ctx.Err())
+		}
+		if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer), grpc.WaitForReady(true)); err != nil {
+			t.Logf("RPC failed, possibly during switchover: %v", err)
+			time.Sleep(100 * time.Millisecond)
```

Comment:
We don't do sleeps in our tests unless there is no other way around. Why do we need this sleep?